### ISSUE TITLE
infra v2: migrate traffic-event-roadworks-bz

### ISSUE THREAD

**Issue by clezag** [noi-techpark/opendatahub-bdp-collectors#652]
This is an issue stub. Include information about the data collector here

**Comment by clezag** [comment]
@rcavaliere in preparation for https://github.com/noi-techpark/bdp-commons/issues/680 , I am migrating this data collector (events origin=PROVINCE_BZ) before making the requested modifications.

Unfortunately, the way in which the event UUID / eventseries UUID are generated, is virtually impossible to replicate, because it relies on random implementation specific Java behavior.
I've already burned a lot of time on this, and the code gets more and more complex to handle all special cases.
I also suspect that even in the old implementation, it's not consistent between Java or library versions. E.g. the UUIDs I generated today on my machine are different from the ones last week in the office.
I know there have been issues with duplicate events in the past. If they are not fixed yet, it could have been caused by this.

To make a long story short: I'd like to change the UUID generation to something more sane and reproducible, and then we have to migrate the old events so that there are no duplicates by hand (using start/end dates, locations and the like).

Do you know of any users who could be impacted by this?

**Comment by rcavaliere** [comment]
@clezag no, nobody is using the data now as far as I know. Yes, make a proposal for a more sustainable solution on how we generate this IDs. I had lot of discussion with Patrick B. at the time on that topic, and I remember that I was not satisfied in the way how he implemented this. The important thing is that the IDs are of course somehow related to the specific information of the event, so that in case of a light update of the event (e.g. traffic jam extending its length) we create a new version (= new record) of the already stored event, without creating a new event.

**Comment by clezag** [comment]
@rcavaliere I would maintain the current mechanism, just simplify the implementation.

Currently it creates a JSON containing `{beginDate, endDate, X, Y}` and then hashes it to generate the UUID.

So every time the start/end/x/y change, a new UUID is generated, and all events that have these fields equal are considered the same event for the purposes of update vs insert.

This will remain all as it is.

The only thing I will change is an internal implementation one: The dates currently are not simple strings, they are complex Java objects with unnecessary internal fields, and the order in which the fields appear in the JSON is nondeterministic. I've implemented the whole structure, just to find out that when I run it on a different day/PC it comes out different and all my tests failed.
```json
{"beginDate":{"year":2025,"month":"JANUARY","era":"CE","dayOfYear":8,"dayOfWeek":"WEDNESDAY","leapYear":false,"dayOfMonth":8,"monthValue":1,"chronology":{"calendarType":"iso8601","id":"ISO"}},"endDate":{"year":2025,"month":"MARCH","era":"CE","dayOfYear":73,"dayOfWeek":"FRIDAY","leapYear":false,"dayOfMonth":14,"monthValue":3,"chronology":{"calendarType":"iso8601","id":"ISO"}},"X":11.3219414845014,"Y":46.4027573679233
```
My proposal is to make these date fields strings, and have the JSON have the fields in a fixed order.
This way it does the same thing, but is better defined and easier to migrate in the future.

Now would be the time to add/remove fields or change the mechanism if you have any feedback regarding that. 
For reference this is the endpoint if you want to look at some raw data: https://www.provinz.bz.it/vmz/traffic.json

**Comment by clezag** [comment]
Migrated datacollector:
https://mobility.api.dev.testingmachine.eu/flat,event/PROVINCE_BZ

Old testing: 
https://mobility.api.opendatahub.testingmachine.eu/v2/flat,event/PROVINCE_BZ

**Comment by clezag** [comment]
@rcavaliere I've now deployed the migrated data collector to testing.
I'm fairly confident that it behaves like the old one (sans UUID generation).
One issue I've seen that we can also fix with the next modifications is the description that's always " | "


To clean up the resulting duplicates (event start, event end, X, Y), I've deleted all but the most recent ones using this query (just documenting it here for when we go to production):

```sql
create table event_bkp_20240102 as select * from event;
delete from event where id in (
	select id from (
		select 
			e.*,
			m.json,
			l.geometry,
			count(1) over (partition by event_interval, l.geometry) as cnt,
			rank() over (partition by event_interval, l.geometry order by e.created_on desc) as rnk,
			e.created_on
		from event e
			join location l on l.id = e.location_id
			join metadata m on m.id = e.meta_data_id
		where origin = 'PROVINCE_BZ'
	) t
	where t.rnk > 1
);
```
